bluemira/equilibria/flux_surfaces.py

```diff
@@ -872,3 +876,251 @@ def poloidal_angle(Bp_strike, Bt_strike, gamma):
+        approximation is valid
+    """
+    # Spherical coords
+    r_f = np.sqrt(x_f**2 + z_f**2)
+    theta_f = np.arctan2(x_f, z_f)
+    # Maxmimum r value for the sphere whithin which harmonics apply
+    max_valid_r = np.amin(r_f)
+
+    # [number of degrees, number of coils]
+    currents2harmonics = np.zeros([max_degree, np.size(r_f)])
+    # First 'harmonic' is constant (this line avoids Nan isuues)
+    currents2harmonics[0, :] = 1  #
+
+    # SH coeffcients from fuction of the current distribution
+    # outside of the sphere coitaining the LCFS
+    # SH coeffs = currents2harmonics @ coil currents
+    for degree in np.arange(1, max_degree):
+        currents2harmonics[degree, :] = (
+            0.5
+            * MU_0
+            * (r_t / r_f) ** degree
+            * np.sin(theta_f)
+            * lpmv(1, degree, np.cos(theta_f))
+            / np.sqrt(degree * (degree + 1))
+        )
+
+    return currents2harmonics, max_valid_r
+
+
+def harmonic_amplitude_marix(
+    collocation_r, collocation_theta, n_collocation, max_degree, r_t
+):
+    """
+    Construct matrix from harmonic amplitudes at given points (in spherical coords).
+
+    The matrix is used in a spherical harmonic approximation of the vacuum/coil
+    contribution to the poilodal flux (psi):
+
+        psi = SUM(
+            A_l * ( r**(l+1) / r_t**l ) * sin (theta) *
+            ( P_l * cos(theta_f) / sqrt(l*(l+1)) )
+        )
+
+    Where l = degree, A_l are the spherical harmonic coeffcients/ampletudes,
+    and is P_l * cos(theta_f) are the associated Legendre polynomials of
+    degree l and order (m) = 1.
+
+    N.B. Vacuum Psi = Total Psi - Plasma Psi.
+
+    Parameters
+    ----------
+    collocation_r: np.array
+        R values of collocation points
+    collocation_theta: np.array
+        Theta values of collocation points
+    n_collocation: integer
+        Number of collocation points
+    max_degree: integer
+        Maximum degree of harmonic to calculate up to
+    r_t: float
+        Typical length scale (e.g. radius at outer midplane)
+
+    Returns
+    -------
+    harmonics2collocation: np.array
+        Matrix of harmonic amplitudes (to get spherical harmonic coefficents
+        use matrix @ coefficents = vector psi_vacuum at colocation points)
+    """
+    # [number of points, number of degrees]
+    harmonics2collocation = np.zeros([n_collocation, max_degree])
+    # First 'harmonic' is constant (this line avoids Nan isuues)
+    harmonics2collocation[:, 0] = 1
+
+    # SH coeffcient matrix
+    # SH coeffs = harmonics2collocation \ vector psi_vacuum at colocation points
+    for degree in np.arange(1, max_degree):
+        harmonics2collocation[:, degree] = (
+            collocation_r ** (degree + 1)
+            * np.sin(collocation_theta)
+            * lpmv(1, degree, np.cos(collocation_theta))
+            / ((r_t**degree) * np.sqrt(degree * (degree + 1)))
+        )
```

Comment:
I'm not sure, but this looks like a sequence of Legendre polynomials. This is quite common: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lpmn.html#scipy.special.lpmn

It looks like you might be able to put this into a vector operation.

The above function can also return the derivatives, which might be useful for optimisation purposes. Perhaps this not quite the equivalent to what you have though.

Reply:
I changed as suggested (function is located in harmonics rather than flux surfaces)